tests/topotests/lib/bgp.py

```diff
@@ -1958,250 +2212,887 @@ def verify_bgp_rib(tgen, addr_type, dut, input_dict, next_hop=None, aspath=None)
+                " o/p {}".format(dut, neighbor_ip)
+            )
+
+            show_bgp_graceful_json = run_frr_cmd(
+                rnode,
+                "show bgp {} neighbor {} graceful-restart json".format(
+                    addr_type, neighbor_ip
+                ),
+                isjson=True,
+            )
+
+            show_bgp_graceful_json_out = show_bgp_graceful_json[neighbor_ip]
+            if show_bgp_graceful_json_out["neighborAddr"] == neighbor_ip:
+                logger.info(
+                    "[DUT: {}]: Neighbor ip matched  {}".format(dut, neighbor_ip)
+                )
+            else:
+                errormsg = "[DUT: {}]: Neighbor ip is NOT matched {}".format(
+                    dut, neighbor_ip
+                )
+                return errormsg
+
+            # Graceful-restart timer
+            if "graceful-restart" in input_dict[peer]["bgp"]:
+                if "timer" in input_dict[peer]["bgp"]["graceful-restart"]:
+                    for rs_timer, value in input_dict[peer]["bgp"]["graceful-restart"][
+                        "timer"
+                    ].items():
+                        if rs_timer == "restart-time":
+
+                            receivedTimer = value
+                            if (
+                                show_bgp_graceful_json_out["timers"][
+                                    "receivedRestartTimer"
+                                ]
+                                == receivedTimer
+                            ):
+                                logger.info(
+                                    "receivedRestartTimer is {}"
+                                    " on {} from peer {}".format(
+                                        receivedTimer, router, peer
+                                    )
+                                )
+                            else:
+                                errormsg = (
+                                    "receivedRestartTimer is not"
+                                    " as expected {}".format(receivedTimer)
+                                )
+                                return errormsg
+
+    logger.debug("Exiting lib API: verify_graceful_restart_timers")
     return True
+
+
+@retry(attempts=4, wait=2, return_is_str=True, initial_wait=2)
+def verify_gr_address_family(tgen, topo, addr_type, addr_family, dut):
+    """
+    This API is to verify gr_address_family in the BGP gr capability advertised
+    by the neighbor router
+
+    Parameters
+    ----------
+    * `tgen`: topogen object
+    * `topo`: input json file data
+    * `addr_type` : ip type ipv4/ipv6
+    * `addr_type` : ip type IPV4 Unicast/IPV6 Unicast
+    * `dut`: input dut router name
+
+    Usage
+    -----
+
+    result = verify_gr_address_family(tgen, topo, "ipv4", "ipv4Unicast", "r1")
+
+    Returns
+    -------
+    errormsg(str) or True
+    """
+
+    logger.debug("Entering lib API: verify_gr_address_family()")
```

Comment:
I think that's error-prone when duplicating a code, maybe we can use everywhere `sys._getframe().f_code.co_name` to print which function is being called?

Reply:
Thanks for comment, I have updated the code.